skiko/src/awtMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt

```diff
@@ -23,9 +23,12 @@ internal class Direct3DRedrawer(
 
     init {
         val adapter = chooseAdapter(properties.adapterPriority.ordinal)
+        if (adapter == 0L) {
+            throw RenderException("Failed to choose DirectX12 adapter.")
+        }
         adapterName = getAdapterName(adapter)
         adapterMemorySize = getAdapterMemorySize(adapter)
-        onDeviceChosen(getAdapterName(adapter))
+        onDeviceChosen(adapterName)
```

Comment:
So, we protected from crash in getAdapterName(0)
But now we have another one crash:
```
if (adapter == 0L) {
            throw RenderException("Failed to choose DirectX12 adapter.")
        }
```

with same condition adapter == 0

Or exception RenderException will handled in another place?

Reply:
Yes, the logic catching RenderException is inside SkiaLayer. When we catch it, we fallback to the next render api.

Windows has this chain of API's: DirectX -> OpenGL -> Software Fast (via native code) -> Software Compat (via Swing)